File: streamer/src/data_static.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

extern unsigned char static_style_volume_css[];
extern unsigned char static_script_volume_js[];
extern unsigned char static_favicon_ico[];
extern unsigned char static_favicon152_png[];
extern unsigned char static_favicon180_png[];
extern unsigned int static_style_volume_css_len;
extern unsigned int static_script_volume_js_len;
extern unsigned int static_favicon_ico_len;
extern unsigned int static_favicon152_png_len;
extern unsigned int static_favicon180_png_len;
unsigned char *data;
unsigned int data_len;

int select_data(char *url, char *rsp) {
  size_t hdr_end;
  if (!strcmp("/favicon.ico", url)) {
    data = static_favicon_ico;
    data_len = static_favicon_ico_len;
    sprintf(rsp, "%s\r\n%s\r\n%s\r\n", "HTTP/1.1 200 OK",
            "Content-Type: image/x-icon",
            "Cache-control: max-age=31536000, immutable");
  } else if (!strcmp("/apple-touch-icon-precomposed.png", url)) {
    data = static_favicon152_png;
    data_len = static_favicon152_png_len;
    sprintf(rsp, "%s\r\n%s\r\n%s\r\n", "HTTP/1.1 200 OK",
            "Content-Type: image/png",
            "Cache-control: max-age=31536000, immutable");
  } else if (!strcmp("/apple-touch-icon.png", url)) {
    data = static_favicon180_png;
    data_len = static_favicon180_png_len;
    sprintf(rsp, "%s\r\n%s\r\n%s\r\n", "HTTP/1.1 200 OK",
            "Content-Type: image/png",
            "Cache-control: max-age=31536000, immutable");
  } else if (!strcmp("/style_volume.css", url)) {
    data = static_style_volume_css;
    data_len = static_style_volume_css_len;
    sprintf(rsp, "%s\r\n%s\r\n%s\r\n", "HTTP/1.1 200 OK",
            "Content-Type: text/css; charset=utf-8",
            "Cache-control: max-age=31536000, immutable");
  } else if (!strcmp("/script_volume.js", url)) {
    data = static_script_volume_js;
    data_len = static_script_volume_js_len;
    sprintf(rsp, "%s\r\n%s\r\n%s\r\n", "HTTP/1.1 200 OK",
            "Content-Type: text/javascript; charset=utf-8",
            "Cache-control: max-age=31536000, immutable");
  } else
    return 1;
  hdr_end = strlen(rsp);
  sprintf(rsp + hdr_end, "%s%u\r\n%s\r\n\r\n", "Content-Length: ", data_len,
          "X-Content-Type-Options: nosniff");
  return 0;
}

int main(int prm_n, char *prm[]) {
  int sock = strtol(prm[1], NULL, 10);
  ssize_t write_size;
  char *rsp = malloc(getpagesize());
  if (select_data(prm[2], rsp))
    execl(resp_err, "resp_err", prm[1], NULL);
  write_size = write(sock, rsp, strlen(rsp));
  write_size += write(sock, data, data_len);
  if (write_size == strlen(rsp) + data_len)
    return 0;
  else
    return 1;
}

File: streamer/rust/data_static.rs
use proxy;
use std::collections::HashMap;
use std::io::Write;
use BufWriter;
use TcpStream;

struct StaticData {
    content_type: String,
    bytes: Vec<u8>,
}

fn data_map() -> HashMap<String, StaticData> {
    let mut data = HashMap::new();
    data.insert(
        "/home.css".to_string(),
        StaticData {
            content_type: "text/css; charset=utf-8".to_string(),
            bytes: include_bytes!("../style/home.css").to_vec(),
        },
    );
    data
}

pub fn send_static(url: &str, req: &Vec<String>, mut stream: BufWriter<TcpStream>) {
    let static_data = data_map();
    match static_data.get(url) {
        Some(resourse) => {
            let hdr = format!(
                "\
HTTP/1.1 200 OK\r\n\
X-Content-Type-Options: nosniff\r\n\
Cache-control: max-age=31536000, immutable\r\n\
Content-Type: {}\r\n\
Content-Length: {}\r\n\r\n",
                resourse.content_type,
                resourse.bytes.len()
            );
            match stream.write_all(hdr.as_bytes()) {
                Ok(_) => match stream.write_all(&resourse.bytes) {
                    _ => (),
                },
                Err(_) => return,
            }
        }
        None => proxy::forward(req.join("\r\n"), stream),
    }
}
